### PR TITLE
Buzzex integration

### PR DESCRIPTION
**Pull request by mrcrislainez**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
New Exchange: Buzzex

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### CODE REVIEW

**Comment by mrcrislainez** [comment]
@askmike, are you able to review this? Thanks.

**Comment by askmike** [comment]
follow up of #2726.

Please don't make a new PR next time, simply add commits that fixes issues and rebase the branch. Thanks!

File: exchange/wrappers/buzzex.js [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+const recoverableErrors = [
+  'SOCKETTIMEDOUT',
+  'TIMEDOUT',
+  'CONNRESET',
+  'CONNREFUSED',
+  'NOTFOUND',
+  'Service:Unavailable',
+  'Request timed out',
+  'Empty response',
+  'API:Invalid nonce',
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: This is exactly the same as what Kraken sends. Unless this exchange actually replies with that error, can you take it out?

File: exchange/wrappers/buzzex.js [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+  );
+}
+
+const recoverableErrors = [
+  'SOCKETTIMEDOUT',
+  'TIMEDOUT',
+  'CONNRESET',
+  'CONNREFUSED',
+  'NOTFOUND',
+  'Service:Unavailable',
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: same as above.

File: exchange/wrappers/buzzex.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  'Request timed out',
+  'Empty response',
+  'API:Invalid nonce',
+  'General:Temporary lockout',
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: same as above.

File: exchange/wrappers/buzzex.js [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  'Response code 525',
+  'Service:Busy'
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: same as above.

File: exchange/wrappers/buzzex.js [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+  const fetch = cb => this.buzzex.api('getinfo', {}, this.handleResponse('getPortfolio', cb, true));
+  retry(null, fetch, handle);
+};
+
+// This assumes that only limit orders are being placed with standard assets pairs
+// It does not take into account volume discounts.
+// Base maker fee is 0.16%, taker fee is 0.26%.
+Trader.prototype.getFee = function(callback) {
+  const makerFee = 0.16;
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: These are the kraken fees, not the Buzzex fees according to: https://support.buzzex.io/hc/en-us/articles/360008709594-What-Is-The-Fee-Structure-On-Buzzex-

File: exchange/wrappers/buzzex.js [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+Trader.prototype.getOrder = function(order, callback) {
+  const handle = (err, data) => {
+    if(err) return callback(err);
+
+    const price = parseFloat( data.orderInfo[ order ].rate );
+    const amount = parseFloat( data.orderInfo[ order ].amount );
+    const date = moment.unix( data.orderInfo[ order ].timestamp_created );
+
+    // TODO: figure out fees, kraken is reporting 0 fees:
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: This applies to kraken, not buzzex.

File: exchange/wrappers/buzzex.js [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+
+Trader.getCapabilities = function () {
+  return {
+    name: 'Buzzex',
+    slug: 'Buzzex',
+    currencies: marketData.currencies,
+    assets: marketData.assets,
+    markets: marketData.markets,
+    requires: ['key', 'secret'],
+    providesHistory: 'date',
```
Comment (askmike) [inline exchange/wrappers/buzzex.js:279]: according to https://api.buzzex.io/docs/#trades there is a tid.

**Review by askmike — CHANGES_REQUESTED** [review]
Thanks for the PR. It's fine if you copy another wrapper (like Kraken's) as an example, but make sure you update all the relevant bits so they apply to Buzzex. Thanks!

**Comment by askmike** [comment]
If you have any questions about the exchange interface as described here: https://gekko.wizb.it/docs/extending/add_an_exchange.html feel free to ask them here, on the forums or reach out to my email (on the bottom of that page) :)

**Comment by mrcrislainez** [comment]
@askmike, i have made the necessary changes.

File: exchange/wrappers/buzzex.js [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+  this.interval = 3100;
+
+  this.buzzex = new Buzzex(
+    this.key,
+    this.secret,
+    {timeout: +moment.duration(60, 'seconds')}
+  );
+}
+
+const recoverableErrors = [
```
Comment (askmike) [inline exchange/wrappers/buzzex.js:37]: can you remove this?

File: exchange/wrappers/buzzex.js [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+Trader.prototype.handleResponse = function(funcName, callback, nonMutating, payload) {
+  return (error, body) => {
+
+    if(!error && !body) {
+      error = new Error('Empty response');
+    }
+
+    if(error) {
+      if(includes(error.message, recoverableErrors)) {
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: this is dead code now.

File: exchange/wrappers/buzzex.js [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        error.notFatal = true;
+      }
+
+      if(includes(error.message, ['Rate limit exceeded'])) {
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: question: Is the Buzzex rate limit message exactly the same as the Kraken rate limit message?

File: exchange/wrappers/buzzex.js [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+        error.notFatal = true;
+        error.backoffDelay = 2500;
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: This is kraken specific due to their API being very slow, is this the case as well for Buzzex? Else 500ms would be a more sane default I think.

File: exchange/wrappers/buzzex.js [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+Trader.prototype.getTrades = function(since, callback, descending) {
+  const startTs = since ? moment(since).valueOf() : null;
+
+  const handle = (err, trades) => {
+
+    if (err) return callback(err);
+
+    var parsedTrades = [];
+    _.each(trades[this.pair], function(trade) {
+      // Even when you supply 'since' you can still get more trades than you asked for, it needs to be filtered
```
Comment (askmike) [inline exchange/wrappers/buzzex.js:81]: Question: Buzzex behaves the same as Kraken here as well?

File: exchange/wrappers/buzzex.js [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+    return callback(undefined, portfolio);
+  };
+
+  const fetch = cb => this.buzzex.api('getinfo', {}, this.handleResponse('getPortfolio', cb, true));
+  retry(null, fetch, handle);
+};
+
+// This assumes that only limit orders are being placed with standard assets pairs
+// It does not take into account volume discounts.
+// Base maker fee is 0.16%, taker fee is 0.26%.
```
Comment (askmike) [inline exchange/wrappers/buzzex.js]: Please update these.

File: exchange/wrappers/buzzex.js [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+  price = this.roundPrice(price); // only round price, not amount
+
+  const handle = (err, data) => {
+    if(err) {
+      return callback(err);
+    }
+
+    let txid;
+
+    if(data.catched) {
```
Comment (askmike) [inline exchange/wrappers/buzzex.js:179]: This is part of an exception flow for Kraken, see here: https://github.com/askmike/gekko/blob/8b3e696ed96cfbc97335d117f27886c77c0d19ea/exchange/wrappers/kraken.js#L133

You have removed the first part meaning that this is dead code.

**Review by askmike — CHANGES_REQUESTED** [review]
Some more nits.

The issue here is that Kraken's API is very hard to work with, and Gekko has a lot of exception handling code to make sure everything stays working. I don't know the Buzzex API, but I'm not assuming it behaves exactly the same way.

Please go through the full file and remove everything kraken specific. Don't only fix the lines I commented on... It's better to NOT handle buzzex errors than to leave old Kraken code that handles errors Buzzex never responds with (this way we can easily add error handling for Buzzex).

Thanks!

File: exchange/wrappers/buzzex.js (hunk 10 above)
Comment (mrcrislainez) [reply exchange/wrappers/buzzex.js]: > Some more nits.
> 
> The issue here is that Kraken's API is very hard to work with, and Gekko has a lot of exception handling code to make sure everything stays working. I don't know the Buzzex API, but I'm not assuming it behaves exactly the same way.
> 
> Please go through the full file and remove everything kraken specific. Don't only fix the lines I commented on... It's better to NOT handle buzzex errors than to leave old Kraken code that handles errors Buzzex never responds with (this way we can easily add error handling for Buzzex).
> 
> Thanks!

Thanks for this. If it's ok to leave error handling at this stage, ill remove codes for it for now. I need to confirm with other devs how's the error messaging of buzzex. Will removing all the codes inside `if(error)` do?

File: exchange/wrappers/buzzex.js (hunk 12 above)
Comment (mrcrislainez) [reply exchange/wrappers/buzzex.js:81]: Yes

**Comment by mrcrislainez** [comment]
@askmike, i have removed the error handling for now. Hopefully this is ok for now, as we are looking to have our internal testing of the Exchange next week, then we also have some testing with Gekko aswell. Thanks.

**Comment by dachen0** [comment]
@mrcrislainez Do you know the api calls of your exchange? If you don't, you might want to look that up instead of using default kraken calls. BTW, you might want to include the api in your package.json lol.

**Comment by askmike** [comment]
@RealAwesomeness you're right, also there are still many references to Kraken's implementation. I'm closing this for now, we can reopen once issues are addressed.

**Pull request closed without merging**